src/meta/src/barrier/mod.rs

```diff
@@ -185,19 +185,23 @@ where
     /// the commands that requires a checkpoint, else we will finish all the commands.
     ///
     /// Returns whether there are still remaining stashed commands to finish.
-    fn finish_commands(&mut self, checkpoint: bool) -> bool {
+    async fn finish_commands(&mut self, checkpoint: bool) -> MetaResult<bool> {
         if checkpoint {
-            self.finished_commands
-                .drain(..)
-                .flat_map(|c| c.notifiers)
-                .for_each(Notifier::notify_finished);
+            for command in self.finished_commands.drain(..) {
```

Comment:
How about moving this out of `if`, in case there're some commands with `checkpoint = false` to finish?